minimal/core_01_minimal.js

```diff
@@ -0,0 +1,258 @@
+  this.moveTo(user.location);
+};
+$.thing.drop.dobj = 'this';
+
+
+// Prototype room: $.room
+$.room = Object.create($.physical);
+$.room.name = 'Prototype room';
+$.room.announce = function(text) {
+  var contents = this.getContents();
+  for (var i = 0; i < contents.length; i++) {
+    var thing = contents[i];
+    if (thing !== user && thing.tell) {
+      thing.tell(text);
+    }
+  }
+};
+
+
+// Prototype user: $.user
+$.user = Object.create($.physical);
+$.user.name = 'Prototype user';
+$.user.connection = null;
+$.user.say = function(text) {
+  user.tell('You say: ' + text);
+  if (user.location && user.location.announce) {
+    user.location.announce(user.name + ' says: ' + text);
+  }
+};
+$.user.say.dobj = 'any';
+$.user.eval = function(code) {
+  user.tell(eval(code));
+};
+$.user.eval.dobj = 'any';
+$.user.tell = function(text) {
+  if (this.connection) {
+    this.connection.write(text);
+  }
+};
+$.user.quit = function() {
+  if (this.connection) {
+    this.connection.close();
+  }
+};
+$.user.quit.dobj = 'none';
+
+
+// Command parser.
+$.execute = function(command) {
+  var argstr = command.trim();
+  var verbstr = argstr;
+  var dobjstr = '';
+  var dobj = null;
+  var space = command.indexOf(' ');
+  if (space !== -1) {
+    verbstr = argstr.substring(0, space).trim();
+    dobjstr = argstr.substring(space).trim();
+  }
+  if (!verbstr) {
+    return;
+  }
+  if (dobjstr) {
+    if (dobjstr === 'me') {
+      dobj = user;
+    } else if (dobjstr === 'here') {
+      dobj = user.location;
+    } else {
+      var objects = [user].concat(user.getContents());
+      if (user.location && user.location.getContents) {
+        objects.push(user.location);
+        objects = objects.concat(user.location.getContents());
+      }
+      for (var i = 0; i < objects.length; i++) {
+        var obj = objects[i];
+        if (obj.name &&
+            obj.name.toLowerCase().indexOf(dobjstr.toLowerCase()) === 0) {
+          dobj = obj;
+          break;
+        }
+      }
+    }
+  }
+  // Collect all objects which could host the verb.
+  var hosts = [user, user.location, dobj];
+  for (var i = 0; i < hosts.length; i++) {
+    var host = hosts[i];
+    if (!host) {
+      continue;
+    }
+    // Check every verb on each object for a match.
+    for (var prop in host) {
+      var func = host[prop];
+      if (prop === verbstr && typeof func === 'function' && func.dobj) {
+        if (func.dobj === 'any' || (func.dobj === 'this' && dobj === host) ||
+            (func.dobj === 'none' && !dobj)) {
+          return host[prop](dobjstr);
+        }
+      }
+    }
+  }
+  user.tell('Command not understood.');
+};
+
+
+$.playerDatabase = Object.create(null);
+
+
+$.connection = {};
+$.connection.onConnect = function() {
+  this.user = null;
+  this.write('Welcome.  Type name of user to connect as (Alpha or Beta).');
+};
+$.connection.onReceive = function(text) {
+  if (this.user) {
+    user = this.user;
+    $.execute(text);
```

Comment:
There is no guarantee that calls to onRecieve will correspond with '\n's.  Socket read boundaries are arbitrary.